compiler/rustc_abi/src/extern_abi.rs

```diff
@@ -249,6 +249,47 @@ impl ExternAbi {
             _ => false,
         }
     }
+
+    /// Can a function with this ABI be called with a rust call expression?
+    ///
+    /// Some ABIs cannot be called from rust, either because rust does not know how to
+    /// generate code for the call, or because a call does not semantically make sense.
+    pub fn can_be_called_with_call_expr(self) -> bool {
+        match self {
+            // Rust doesn't know how to call functions with this ABI.
+            ExternAbi::Custom => false,
+
+            // These define entry points for the host, and cannot be called on the GPU.
+            ExternAbi::PtxKernel | ExternAbi::GpuKernel => false,
+
+            // The interrupt ABIs should only be called by the CPU.
+            // They have complex pre- and postconditions.
+            ExternAbi::AvrInterrupt
+            | ExternAbi::AvrNonBlockingInterrupt
+            | ExternAbi::Msp430Interrupt
+            | ExternAbi::RiscvInterruptM
+            | ExternAbi::RiscvInterruptS
+            | ExternAbi::X86Interrupt => false,
+
+            ExternAbi::C { .. }
+            | ExternAbi::System { .. }
+            | ExternAbi::Rust
+            | ExternAbi::RustCall
+            | ExternAbi::RustCold
+            | ExternAbi::Unadjusted
+            | ExternAbi::EfiApi
+            | ExternAbi::Aapcs { .. }
+            | ExternAbi::CCmseNonSecureCall
+            | ExternAbi::CCmseNonSecureEntry
+            | ExternAbi::Cdecl { .. }
+            | ExternAbi::Stdcall { .. }
+            | ExternAbi::Fastcall { .. }
+            | ExternAbi::Thiscall { .. }
+            | ExternAbi::Vectorcall { .. }
+            | ExternAbi::SysV64 { .. }
+            | ExternAbi::Win64 { .. } => true,
+        }
```

Comment:
Asking this about all `ExternAbi`, instead of performing the analysis at the site and using a `match` on `CanonAbi` which allows handling the `Interrupt` variant which includes all its nested variants, and also matches directly with whether we can generate code, seems... less than ideal. We should be rejecting anything that has a nonsensical mapping in `rustc_ast_lowering` after https://github.com/rust-lang/rust/pull/142134 lands, and hitting `AbiMapping::Invalid` after trying to `canonize_abi` for the case suggests something has gone direly wrong and we are fine if we ICE on it.

Reply:
So, do you want me to do that now? Or wait? Is there a better way of getting at a `CanonAbi` than

```rust
        match AbiMap::from_target(&self.sess().target).canonize_abi(abi, false) {
            AbiMapping::Direct(canon_abi) | AbiMapping::Deprecated(canon_abi) => {
            }
        }
```

Reply:
Right now, there is not!

Probably should shove it in a field on Target at least.

Reply:
it's still not exactly clear to me what you want to see happen to this PR (versus what changes would be good to make down the line)

Reply:
Allright, I'm using `AbiMap` now, and it's definitely an improvement already.